neqo-transport/src/connection/mod.rs

```diff
@@ -495,36 +496,53 @@ impl Connection {
         Ok(())
     }
 
-    /// Access the latest resumption token on the connection.
-    pub fn resumption_token(&mut self) -> Option<ResumptionToken> {
-        if self.state < State::Connected {
-            return None;
+    fn create_resumption_token(&mut self, now: Instant) {
+        if self.role == Role::Server
+            || self.state < State::Connected
+            || !self.crypto.has_resumption_token()
+        {
+            return;
         }
-        match self.crypto.tls {
-            Agent::Client(ref mut c) => match c.resumption_token() {
-                Some(ref t) => {
-                    qtrace!("TLS token {}", hex(t.as_ref()));
-                    let mut enc = Encoder::default();
-                    let rtt = self.loss_recovery.rtt();
-                    let rtt = u64::try_from(rtt.as_millis()).unwrap_or(0);
-                    enc.encode_varint(rtt);
-                    enc.encode_vvec_with(|enc_inner| {
-                        self.tps
-                            .borrow()
-                            .remote
-                            .as_ref()
-                            .expect("should have transport parameters")
-                            .encode(enc_inner);
-                    });
-                    let token = self.new_token.take_token();
-                    enc.encode_vvec(token.as_ref().map_or(&[], |t| &t[..]));
-                    enc.encode(t.as_ref());
-                    qinfo!("resumption token {}", hex_snip_middle(&enc[..]));
-                    Some(ResumptionToken::new(enc.into(), t.expiration_time()))
-                }
-                None => None,
-            },
-            Agent::Server(_) => None,
+
+        // If a NEW_TOKEN frame has not be received yet wait for it for a short time.
+        if !self.new_token.has_token() {
+            if self.release_resumption_token_timer.is_none() {
+                self.release_resumption_token_timer =
+                    Some(now + 3 * self.loss_recovery.pto_raw(PNSpace::ApplicationData));
+                return;
+            }
+            if self.release_resumption_token_timer.unwrap() > now {
+                return;
+            }
+        }
+
+        // Create reumpion tokens:
+        //  1) create only one if we have not receiveed a new_token,
+        //  2) otherwise create as many as possible that contain ticket and token.
+        loop {
```

Comment:
A few things:

The crypto code shouldn't know about the events.  It can create a token without dispatching it.

You duplicate the timer update code, but that is the same in both places; I would duplicate the token generation piece instead.

That suggests a different structure:

```rust
if self.role == Role::Server || self.state < State::Connected {
    return;
}
while self.crypto.has_resumption_token() && self.new_token.has_token() {
    let tok = self.crypto.create_resumption_token(
        &self.new_token.take_token(),
        self.tps.borrow().remote.as_ref().expect("..."),
        u64::try_from(self.loss_recovery.rtt().as_millis()).unwrap_or(0)),
    );
    self.events.resumption_token(tok);
}
if self.crypto.has_resumption_token() {
    let arm = if let Some(t) = self.resumption_token_timer {
        if t < now {
            let tok = self.crypto.create_resumption_token(
                &[],
                self.tps.borrow().remote.as_ref().expect("..."),
                u64::try_from(self.loss_recovery.rtt().as_millis()).unwrap_or(0)),
            );
            self.events.resumption_token(tok);
            // This means that we release one session ticket every 3 PTOs
            // if no NEW_TOKEN frame is received.
            self.crypto.has_resumption_token()
        } else {
            false
        }
    } else {
        true
    };
    self.resumption_token_timer = if arm {
        Some(3 * self.loss_recovery.pto_raw(PNSpace::ApplicationData))
    } else {
        None
    };
}
```

I considered the option of moving this elsewhere, but this seems like the right trade-off.

A different design would have a partial event created for each session ticket that we receive.  That would be withheld unless that event was too old or a NEW_TOKEN completed it.  However, that means that next_event needs to know what time it is.  We also hit next_event more often than we hit process_output.  The cost of this design is that process_output needs to be run in tests, which has odd side effects that we need to manage, but the tests where we use tokens are fairly limited and we can just deal with the side effects.